include/session/network.hpp

```diff
@@ -0,0 +1,421 @@
+    /// paths then new paths will be constructed by opening and testing connections to random
+    /// service nodes in the snode pool.
+    ///
+    /// Inputs:
+    /// - `excluded_node` -- [in, optional] node which should not be included in the path.
+    /// - `callback` -- [in] callback to be triggered once we have a valid path, NULL if we are
+    /// unable to find a valid path.
+    void with_path(
+            std::optional<service_node> excluded_node,
+            std::function<void(std::optional<onion_path> path, std::optional<std::string> error)>
+                    callback);
+
+    /// API: network/build_paths_if_needed
+    ///
+    /// Builds onion request paths if needed by opening and testing connections to random service
+    /// nodes in the snode pool.  If we already have enough paths the callback will be triggered
+    /// with the current paths.
+    ///
+    /// Inputs:
+    /// - `excluded_node` -- [in, optional] node which should not be included in the paths.
+    /// - `callback` -- [in] callback to be triggered once we have enough paths.  NOTE: If we are
+    /// unable to create the paths the callback will be triggered with an empty list.
+    void build_paths_if_needed(
+            std::optional<service_node> excluded_node,
+            std::function<
+                    void(std::vector<onion_path> updated_paths, std::optional<std::string> error)>
+                    callback);
+
+    /// API: network/get_service_nodes_recursive
+    ///
+    /// A recursive function that will attempt to retrieve service nodes from a given node until it
+    /// successfully retrieves nodes or the list is drained.
+    ///
+    /// Inputs:
+    /// - `target_nodes` -- [in] list of nodes to send requests to until we get a result or it's
+    /// drained.
+    /// - `limit` -- [in, optional] the number of service nodes to retrieve.
+    /// - `callback` -- [in] callback to be triggered once we receive nodes.  NOTE: If we drain the
+    /// `target_nodes` and haven't gotten a successful response an empty list will be returned along
+    /// with an error string.
+    void get_service_nodes_recursive(
+            std::vector<service_node> target_nodes,
+            std::optional<int> limit,
+            std::function<void(std::vector<service_node> nodes, std::optional<std::string> error)>
+                    callback);
+
+    /// API: network/find_valid_guard_node_recursive
+    ///
+    /// A recursive function that sends a request to provided nodes until a successful response is
+    /// received or the list is drained.
+    ///
+    /// Inputs:
+    /// - `target_nodes` -- [in] list of nodes to send requests to until we get a result or it's
+    /// drained.
+    /// - `callback` -- [in] callback to be triggered once we make a successful request.  NOTE: If
+    /// we drain the `target_nodes` and haven't gotten a successful response a NULL
+    /// `connection_info` and empty will be provided.
```

Comment:
I think this got copied from the C API, but I'm not sure what it should be for the C++ API.